src/listeners.js

```diff
@@ -0,0 +1,145 @@
+        exchange: 'exchange/taskcluster-hooks/v1/hook-deleted',
+        routingKeyPattern: '#',
+      }],
+      queueName: 'hookChanged',
+      maxLength : 50,
+      //expires : 1800000,
+
+    }, (msg) => this.reconcileConsumers()
+    );
+    debug('Listening to hook exchanges');
+    this.pulseHookChangedListener = consumer;
+    this.listeners = [];
+  }
+
+  /** Create a new pulse consumer for a hook */
+  async createListener({hookId, hookGroupId, bindings}, oldBindings) {
+    const hook = await this.Hook.load({hookGroupId, hookId}, true);
+    if (hook) {
+      let listener = await pulse.consume({
+        client,      
+        queueName: `${hookGroupId}/${hookId}`, // serves as unique id for every listener
+        maxLength : 50,
+      }, async ({payload}) => {
+        console.log(payload);
+        const hook = this.hook;
+        // Pass empty functions in res which are executed by `triggerHookCommon`
+        // This is more of a hack so that if the payload doesnt validate against
+        // `triggerSchema`,git  we can simply ignore the message.
+        const res = {
+          reply: function() {},
+          reportError: function() {},
+        };
+
+        await triggerHookCommon.call(this, {res, hook, payload, firedBy: 'pulseHook'});
+      });
+      for (let {exchange, routingKeyPattern} of oldBindings) {
+        await this.client.withChannel.unbindQueue(queueName, exchange, routingKeyPattern);
+      }
+      for (let {exchange, routingKeyPattern} of bindings) {
+        await this.client.withChannel.bindQueue(queueName, exchange, routingKeyPattern);
+      }
+
+      this.listeners.push(listener);
+    }
+  }
+  
+  async reconcileConsumers() {
+    let Queues = [];
+    let allHooksData = [];
+    console.log('helelo');
+    await this.Queues.scan(
+      {},
+      {
+        limit: 1000,
+        handler:(queue) => Queues.push(queue),
+      }
+    );
+
+    let continuationToken = undefined;
+    
+    do {
+      allHooksData  = await this.Hooks.scan({
+        bindings:   Entity.op.notEqual([]),
+      }, {
+        limit: 1000,
+        continuation: continuationToken,
+      });
+      continuationToken = allHooksData.continuation;
+      allHooksData.entries.forEach(hook => {
+        const {hookGroupId, hookId} = hook;
+        const queue = _.find(Queues, {hookGroupId, hookId});
+        if (queue) {
+          const index = this.listeners.findIndex(({_queueName}) => listener._queueName === queue.queueName);
+          if (index == -1) {
+            this.createListener(hook, queue.bindings);
+          }
+          _.pull(Queues, queue);
+        } else {
+          this.createListener(hook, {});
+        }
+      });
+    } while (continuationToken);
+    
+    // Delete the queues now left in the Queues list.
+    Queues.forEach(async (queue) => {     
+      if (this.client.withChannel.checkQueue(queue.queueName)) {
+        this.client.withChannel.deleteQueue(queue.queueName);
+      }
```

Comment:
I think you'll need to remove the listener from `this.listeners` here, too.

Reply:
This still stands.  And I didn't quite say that right -- you'll need to iterate over the listeners in `this.listeners` and call `listener.stop()` for each one.

Reply:
`this.listeners` also contains the new `listeners` created during reconciling the `Hooks` Azure table with the `Queues` Azure table. So instead of stopping all the listeners in `this.listeners`, we only have to stop the `listeners`( aka `consumers`) of the `queues`( `queueName`=`$(hookGroupId)/$(hookId)` ) which are no longer in the `Hooks` Azure table.

A queue is pulled from the `Queues` list when its corresponding `hook` is found in the `Hooks` table.  So at the end of the `do..while` loop the `Queues` list has all the queues which are no longer of use to us, and must be deleted. and the corresponding listener stopped